File: cpp/pricing/interest/discount.cpp
#pragma once
/*
   Finite-frequency compounding & fixed-rate bond helpers
   ======================================================

   Header-only; include <cmath>, <vector>, <functional>.
   All rates in decimals, times in years.  Deterministic cash-flows, no options.
*/

#include <cmath>
#include <vector>
#include <functional>
#include <stdexcept>
#include <limits>

namespace fc::interest {

// ---------- Elementary transforms -------------------------------------------------

inline double discount_factor(double rate, int m, double t) {
    return std::pow(1.0 + rate / m, -m * t);
}

inline double future_value(double pv, double rate, int m, double t) {
    return pv * std::pow(1.0 + rate / m, m * t);
}

inline double present_value(double fv, double rate, int m, double t) {
    return fv * discount_factor(rate, m, t);
}

inline double effective_annual_rate(double rate, int m) {
    return std::pow(1.0 + rate / m, m) - 1.0;
}

inline double equivalent_cont_rate(double rate, int m) {
    return m * std::log(1.0 + rate / m);
}

// ---------- Cash-flow scaffolding --------------------------------------------------

inline std::vector<std::pair<double, double>>
generate_cashflows(double face, double coupon_rate, int m, double n_years)
{
    const double dt = 1.0 / m;
    const int n_payments = static_cast<int>(std::round(n_years * m));
    const double coupon = coupon_rate * face / m;
    std::vector<std::pair<double, double>> flows;
    flows.reserve(n_payments);

    for (int k = 1; k <= n_payments; ++k)
        flows.emplace_back(k * dt, coupon);

    // add principal to last coupon
    flows.back().second += face;
    return flows;
}

// ---------- Pricing ----------------------------------------------------------------

inline double price_fixed_rate_bond(
    double face,
    double coupon_rate,
    int m,
    double n_years,
    const std::function<double(double)>& curve // zero rate y(t)
)
{
    double pv = 0.0;
    for (auto [t, cf] : generate_cashflows(face, coupon_rate, m, n_years)) {
        const double y = curve(t);
        pv += cf * discount_factor(y, m, t);
    }
    return pv;
}

// ---------- Yield-to-Maturity -------------------------------------------------------

inline double yield_to_maturity(
    double price,
    double face,
    double coupon_rate,
    int m,
    double n_years,
    double tol = 1e-10,
    int    max_iter = 100,
    double guess = 0.03
)
{
    const auto flows = generate_cashflows(face, coupon_rate, m, n_years);

    auto pv_at = [&](double y) {
        double pv = 0.0;
        for (auto [t, cf] : flows)
            pv += cf * discount_factor(y, m, t);
        return pv;
    };

    // Newton iteration
    double y = guess;
    for (int i = 0; i < max_iter; ++i) {
        double pv  = pv_at(y);
        double diff = pv - price;
        if (std::abs(diff) < tol) return y;

        double d_pv = 0.0;
        for (auto [t, cf] : flows) {
            d_pv += -m * t * cf * discount_factor(y, m, t) / (1.0 + y / m);
        }
        double step = diff / d_pv;
        y -= step;
        if (!(y > 0.0 && y < 1.0)) break;        // fall back to bisection
    }

    // Bisection safeguard
    double lo = 0.0, hi = 1.0;
    for (int i = 0; i < max_iter; ++i) {
        double mid = 0.5 * (lo + hi);
        if (pv_at(mid) > price)
            lo = mid;
        else
            hi = mid;
        if (hi - lo < tol) return mid;
    }
    throw std::runtime_error("yield_to_maturity did not converge");
}

// ---------- Risk measures ----------------------------------------------------------

inline double dv01(
    double price,
    double face,
    double coupon_rate,
    int m,
    double n_years,
    double ytm,
    double bp = 1e-4
)
{
    const auto curve_up   = [ytm, bp](double){ return ytm + bp; };
    const auto curve_down = [ytm, bp](double){ return ytm - bp; };

    double pv_up   = price_fixed_rate_bond(face, coupon_rate, m, n_years, curve_up);
    double pv_down = price_fixed_rate_bond(face, coupon_rate, m, n_years, curve_down);

    return 0.5 * (pv_down - pv_up);
}

} // namespace fc::interest

File: python/pricing/interest/discount.py
"""
Finite-frequency compounding & fixed-rate bond helpers
=====================================================

All rates are quoted in decimals (0.05 == 5 %), all times in *years*.
Day-count conversion happens **up-stream**: pass actual year-fractions
into the `t` arguments.

Author: FinCraftr project
"""
from __future__ import annotations

import math
from typing import Iterable, Callable, Sequence, Tuple, List


# ---------- Elementary transforms -------------------------------------------------

def discount_factor(rate: float, m: int, t: float) -> float:
    """(1 + rate/m)^(-m*t)  – finite-frequency discount factor."""
    return (1.0 + rate / m) ** (-m * t)


def future_value(pv: float, rate: float, m: int, t: float) -> float:
    """Compounds present value forward m-times-per-year for t years."""
    return pv * (1.0 + rate / m) ** (m * t)


def present_value(fv: float, rate: float, m: int, t: float) -> float:
    """Discounts a future value back to today."""
    return fv * discount_factor(rate, m, t)


def effective_annual_rate(rate: float, m: int) -> float:
    return (1.0 + rate / m) ** m - 1.0


def equivalent_cont_rate(rate: float, m: int) -> float:
    return m * math.log(1.0 + rate / m)


# ---------- Cash-flow scaffolding --------------------------------------------------

def generate_cashflows(
    face: float,
    coupon_rate: float,
    m: int,
    n_years: float
) -> List[Tuple[float, float]]:
    """
    Returns [(t_1, C_1), ..., (t_N, C_N + redemption)] where time is in years.
    Assumes regular schedule, no stubs, deterministic amounts.
    """
    dt = 1.0 / m
    n_payments = int(round(n_years * m))
    coupon = coupon_rate * face / m
    flows = [(k * dt, coupon) for k in range(1, n_payments + 1)]
    # add principal to last coupon
    t_last, c_last = flows[-1]
    flows[-1] = (t_last, c_last + face)
    return flows


# ---------- Pricing ----------------------------------------------------------------

def price_fixed_rate_bond(
    face: float,
    coupon_rate: float,
    m: int,
    n_years: float,
    curve: Callable[[float], float],
) -> float:
    """
    Present value of a fixed-rate bond under deterministic curve.
    `curve(t)` must return *annualised* zero rate on the **same** compounding basis m.
    """
    pv = 0.0
    for t, cf in generate_cashflows(face, coupon_rate, m, n_years):
        y_t = curve(t)
        pv += cf * discount_factor(y_t, m, t)
    return pv


# ---------- Yield-to-Maturity -------------------------------------------------------

def yield_to_maturity(
    price: float,
    face: float,
    coupon_rate: float,
    m: int,
    n_years: float,
    tol: float = 1e-10,
    max_iter: int = 100,
    guess: float = 0.03,
) -> float:
    """
    Solves PV(price) = price for the single YTM using Newton with bisection fallback.
    Returns nominal annual yield on the same compounding basis m.
    """
    flows = generate_cashflows(face, coupon_rate, m, n_years)

    def pv_at(y: float) -> float:
        return sum(cf * discount_factor(y, m, t) for t, cf in flows)

    # Newton iteration
    y = guess
    for _ in range(max_iter):
        pv = pv_at(y)
        diff = pv - price
        if abs(diff) < tol:
            return y
        # derivative dPV/dy
        d_pv = sum(
            -m * t * cf * discount_factor(y, m, t) / (1 + y / m)
            for t, cf in flows
        )
        step = diff / d_pv
        y -= step
        if not (0.0 < y < 1.0):  # Newton wandered off –> bisection
            break
    # Bisection safeguard
    lo, hi = 0.0, 1.0
    for _ in range(max_iter):
        mid = 0.5 * (lo + hi)
        if pv_at(mid) > price:
            lo = mid
        else:
            hi = mid
        if hi - lo < tol:
            return mid
    raise RuntimeError("YTM solver failed to converge")


# ---------- Risk measures ----------------------------------------------------------

def dv01(price: float, face: float, coupon_rate: float, m: int, n_years: float, ytm: float, bp: float = 1e-4) -> float:
    """Dollar value of a 1bp move, finite diff."""
    y_up = ytm + bp
    y_down = ytm - bp
    pv_up = price_fixed_rate_bond(face, coupon_rate, m, n_years, lambda _: y_up)
    pv_down = price_fixed_rate_bond(face, coupon_rate, m, n_years, lambda _: y_down)
    return 0.5 * (pv_down - pv_up)
